File: relisoft-hr/server/controllers/notificationController.js
import Notification from '../models/Notification.js';

export const getNotifications = async (req, res) => {
  try {
    const { page = 1, limit = 20 } = req.query;

    const notifications = await Notification.find({ user: req.user.id })
      .skip((page - 1) * limit)
      .limit(Number(limit))
      .sort('-createdAt');

    const total = await Notification.countDocuments({ user: req.user.id });

    res.status(200).json({
      success: true,
      data: notifications,
      pagination: {
        page: Number(page),
        limit: Number(limit),
        total,
        pages: Math.ceil(total / limit),
      },
    });
  } catch (error) {
    res.status(500).json({ success: false, message: error.message });
  }
};

export const markRead = async (req, res) => {
  try {
    const notification = await Notification.findByIdAndUpdate(
      req.params.id,
      { readAt: Date.now(), isRead: true },
      { new: true }
    );

    if (!notification) {
      return res.status(404).json({ success: false, message: 'Notification not found' });
    }

    res.status(200).json({ success: true, data: notification });
  } catch (error) {
    res.status(500).json({ success: false, message: error.message });
  }
};

export const markAllRead = async (req, res) => {
  try {
    await Notification.updateMany(
      { user: req.user.id, isRead: false },
      { isRead: true, readAt: Date.now() }
    );

    res.status(200).json({ success: true, message: 'All notifications marked as read' });
  } catch (error) {
    res.status(500).json({ success: false, message: error.message });
  }
};

export const getUnreadCount = async (req, res) => {
  try {
    const count = await Notification.countDocuments({
      user: req.user.id,
      isRead: false,
    });

    res.status(200).json({ success: true, data: { unreadCount: count } });
  } catch (error) {
    res.status(500).json({ success: false, message: error.message });
  }
};

export const createNotification = async (req, res) => {
  try {
    const notification = await Notification.create({
      ...req.body,
      createdBy: req.user.id,
    });

    res.status(201).json({ success: true, data: notification });
  } catch (error) {
    res.status(500).json({ success: false, message: error.message });
  }
};

File: relisoft-hr/server/Controllers/NotificationController.cs
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;
using Microsoft.EntityFrameworkCore;
using RelisoftHR.Data;
using RelisoftHR.DTOs;
using RelisoftHR.Models;

namespace RelisoftHR.Controllers;

[ApiController]
[Route("api/notifications")]
[Authorize]
public class NotificationController : ControllerBase
{
    private readonly AppDbContext _db;
    public NotificationController(AppDbContext db) => _db = db;

    private int GetUserId()
    {
        var claim = User.FindFirst(System.Security.Claims.ClaimTypes.NameIdentifier)?.Value;
        return claim != null ? int.Parse(claim) : 0;
    }

    [HttpGet]
    public async Task<ActionResult> GetNotifications([FromQuery] bool? unreadOnly)
    {
        var empId = GetUserId();
        var query = _db.Notifications.Where(n => n.EmployeeId == empId).AsQueryable();
        if (unreadOnly == true)
            query = query.Where(n => !n.IsRead);
        var list = await query.OrderByDescending(n => n.CreatedOn).ToListAsync();
        return Ok(list);
    }

    [HttpPost]
    public async Task<ActionResult> CreateNotification([FromBody] NotificationRequest req)
    {
        if (!await _db.Employees.AnyAsync(e => e.Id == req.EmployeeId)) return NotFound();
        var notification = new Notification
        {
            EmployeeId = req.EmployeeId, Title = req.Title, Message = req.Message,
            Category = req.Category, Link = req.Link, CreatedOn = DateTime.UtcNow
        };
        _db.Notifications.Add(notification);
        await _db.SaveChangesAsync();
        return Ok(notification);
    }

    [HttpPost("{id}/read")]
    public async Task<ActionResult> MarkRead(int id)
    {
        var n = await _db.Notifications.FindAsync(id);
        if (n == null) return NotFound();
        if (n.EmployeeId != GetUserId()) return Forbid();
        n.IsRead = true;
        await _db.SaveChangesAsync();
        return Ok(n);
    }

    [HttpPost("read-all")]
    public async Task<ActionResult> MarkAllRead()
    {
        var empId = GetUserId();
        var unread = await _db.Notifications.Where(n => n.EmployeeId == empId && !n.IsRead).ToListAsync();
        foreach (var n in unread) n.IsRead = true;
        await _db.SaveChangesAsync();
        return Ok(new { message = "All marked as read" });
    }

    [HttpGet("unread-count")]
    public async Task<ActionResult> GetUnreadCount()
    {
        var empId = GetUserId();
        var count = await _db.Notifications.CountAsync(n => n.EmployeeId == empId && !n.IsRead);
        return Ok(new { count });
    }

    [HttpGet("templates")]
    public async Task<ActionResult> GetTemplates()
    {
        var list = await _db.NotificationTemplates.Where(t => t.IsActive).ToListAsync();
        return Ok(list);
    }

    [HttpPost("templates")]
    public async Task<ActionResult> CreateTemplate([FromBody] NotificationTemplateRequest req)
    {
        var template = new NotificationTemplate
        {
            EventType = req.EventType, Title = req.Title, Message = req.Message, IsActive = req.IsActive
        };
        _db.NotificationTemplates.Add(template);
        await _db.SaveChangesAsync();
        return Ok(template);
    }
}
